### PR TITLE
Add contextual commands

### PR DESCRIPTION
**Pull request by youknowriad**
Related #50407 and #48457

## What?

This PR adds support for contextual commands to the command center by providing:

 - A `setContext` action in the commands store to set the current context.
 - Allowing command loaders and commands to provide a "context" key when registering them. 

The behavior is the following:

 - When you open the command center, the commands of the current context are rendered.
 - When you search the command center, the behavior remains unchanged and any command can be rendered.

For the current PR, I only added a single contextual command: "Browse styles" to the site editor.

## Testing Instructions

1- Open the site editor
2- In view mode, notice that the "browse styles" command is visible as soon as you open the command center
3- In edit mode, notice that the "browse styles" is not visible but is accessible if you search for it.

### CODE REVIEW

**Comment by jameskoster** [comment]
I'll preface my comment with a caveat to ignore it if this is purely a technical implementation for further expansion :)

In terms of contexts, "Browse styles" is a bit lacklustre at the root, especially as there's already a "Styles" menu item. It might be better to concentrate on a different context first. 

If we migrated the current template details popover actions into the command palette we could close https://github.com/WordPress/gutenberg/issues/48481 (after https://github.com/WordPress/gutenberg/pull/50369 is merged too). That would mean: When editing a template, display the "Reset" / "Delete" action, a shortcut to manage templates, and (as a welcome addition) shortcuts to other recently edited templates:

Screenshot 2023-05-12 at 12 40 35

What do you think?

**Comment by youknowriad** [comment]
Yes, that command was just an example that I took from the related issue. I can remove it for now and focus on the the "edit" context instead of the "view" one.

**Comment by youknowriad** [comment]
@jameskoster I've updated the PR, I removed the "browse styles" command and added "clear customization" and "delete" template commands.

Let me know how it feels.

The goal of this PR is not to be exhaustive in terms of available contextual commands but more to prove the API instead. Let me know if it achieves that or not.

**Comment by jameskoster** [comment]
I'm unable to build the PR atm – it seems there's an [unresolved conflict](https://github.com/WordPress/gutenberg/pull/50543/files#diff-ecffee0daf6148e5e4693a18d52efb690aeb1db805f961e0466114da832ae6b6R58-R77), but the description seems good :)

**Comment by youknowriad** [comment]
Oh sorry about that, I had fixed the conflict but forgot to save the file 🙃 It should fine now

File: packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js [hunk 1]
```diff
@@ -32,7 +32,7 @@ export function SidebarNavigationItemGlobalStyles( props ) {
 			<NavigatorButton
 				{ ...props }
 				as={ SidebarNavigationItem }
-				path="/wp_global_styles"
+				path="/styles"
```
Comment (ntsekouras) [inline packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js]: This is not related, right?

File: packages/commands/src/components/command-menu.js [hunk 2]
```diff
@@ -90,17 +90,31 @@ export function CommandMenuLoaderWrapper( { hook, search, setLoader, close } ) {
...
+				loaders: hasSearch
+					? getCommandLoaders( group )
+					: getContextualCommandLoaders( group ),
 			};
 		},
-		[ group ]
+		[ group, hasSearch ]
 	);
 
+	if ( ! commands.length && ! loaders.length ) {
```
Comment (ntsekouras) [inline packages/commands/src/components/command-menu.js:104]: Here we can have loaders but not commands. You can test this by invoking it in a theme template without changes.

File: packages/commands/src/hooks/use-command-context.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+	useEffect( () => {
+		setContext( context );
+	}, [ context, setContext ] );
+
+	// This effects ensures that on unmount, we restore the context
+	// that was set before the component actually mounts.
+	useEffect( () => {
+		const initialContextRef = initialContext.current;
+		return () => setContext( initialContextRef );
```
Comment (ntsekouras) [inline packages/commands/src/hooks/use-command-context.js:30]: Can you explain a bit more about this? If I understand correctly this is not used right now and I'm not sure when the edit-site Layout could unmount.

File: packages/edit-site/src/hooks/commands/use-edit-mode-commands.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+	if ( isRemovable ) {
+		const label =
+			template.type === 'wp_template'
+				? __( 'Delete template' )
+				: __( 'Delete template part' );
+		commands.push( {
+			name: label,
+			label,
+			icon: trash,
+			context: 'site-editor-edit',
```
Comment (ntsekouras) [inline packages/edit-site/src/hooks/commands/use-edit-mode-commands.js:41]: Do you think we will have multiple contexts in the future? For example being in a template with a block selected. That would mean commands for the template, but also for the block(copy/duplicate or whatever).

**Review by ntsekouras — COMMENTED** [review]
Thanks for the PR Riad! Besides my few comments, I found myself expecting these contextual commands to be on the top while searching. This ordering can probably be handled separately though.
Screenshot 2023-05-15 at 3 37 07 PM

File: packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js (hunk 1 above)
Comment (youknowriad) [reply packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js]: Initially I had a "browse styles" command and I wanted the path to be `/styles` as it's better than `/wp_global_styles` (it persists in the url). I can remove from the PR if needed.

File: packages/commands/src/components/command-menu.js (hunk 2 above)
Comment (youknowriad) [reply packages/commands/src/components/command-menu.js:104]: That's normal, but ideally it shouldn't show the empty space it's showing now.

File: packages/commands/src/hooks/use-command-context.js (hunk 3 above)
Comment (youknowriad) [reply packages/commands/src/hooks/use-command-context.js:30]: This is a solution to the problem of third-part contexts. Imagine you have a component that is only rendered in some conditions and that when rendered sets a different third-party context for the command center. The code here ensures that whenever that components unmount, we restore the context that was in the store before that particular component mounts. It's not perfect but it's probably acceptable.

File: packages/edit-site/src/hooks/commands/use-edit-mode-commands.js (hunk 4 above)
Comment (youknowriad) [reply packages/edit-site/src/hooks/commands/use-edit-mode-commands.js:41]: probably yes.

File: packages/commands/src/components/command-menu.js (hunk 2 above)
Comment (youknowriad) [reply packages/commands/src/components/command-menu.js:104]: Solved

**Comment by jameskoster** [comment]
For the clear customisations command:

* Now feels like a good time to consider the copy. "Reset template" might be simpler. If we don't want to do that here, then I'd lean towards keeping the same label, IE "Clear customizations" rather than "Clear template customizations" as it is on the PR.
* I'm not sure the Trash icon is quite right for this command. From the current library, `backup` might be a good fit, and conceptually links the feature to Revisions. 
* If you execute this command elsewhere in the UI a confirmation Snackbar with undo action is displayed. I didn't see that when executing via the command palette.

> I found myself expecting these contextual commands to be on the top while searching

I'd agree with that. They could potentially benefit from some visual separation too:

Screenshot 2023-05-15 at 13 58 58

On that point, and in case it's useful for this PR, the original designs had two groups for contextual suggestions, visually separated like in the screenshot above:

1. Actions, e.g. delete, add new, publish
2. Shortcuts, e.g. links to recently edited entries, and to management views

File: packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js (hunk 1 above)
Comment (ntsekouras) [reply packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js]: No need :)

**Comment by youknowriad** [comment]
I pushed an update to get very close to the mockup and desired behavior above.

**Comment by youknowriad** [comment]
How is this looking, can we merge it?

**Comment by jameskoster** [comment]
I'm still not seeing the Snackbar confirmation when resetting a template via the command palette. This is quite important, because when the customisations are small it's not always obvious whether the action was performed successfully or not.

**Comment by youknowriad** [comment]
I'll take a look, the place where I copied the code from, there was no snackbar.

**Comment by youknowriad** [comment]
Snackbar added back.

**Comment by jameskoster** [comment]
> I'll take a look, the place where I copied the code from, there was no snackbar.

Oh no 🙈  Where did you copy from? I'll make an issue.

This is working as expected now, including the Snackbar 🚀

Screenshot 2023-05-16 at 11 11 03

**Review by jameskoster — APPROVED** [review]

**Comment by youknowriad** [comment]
> Oh no 🙈 Where did you copy from? I'll make an issue.

No need, it seems that one saves the template right away and displays a different snackbar separately. (list view)
